lib/src/impl_ffi/impl_ffi.utils.dart

```diff
@@ -14,40 +14,101 @@
+/// [_finalizerReachabilityFence] is used after usage.
 ///
-/// Callers should be ware that the Dart GC may collect [key] as soon as it
-/// deems the object to not be in use anymore. This can happy at any point when
-/// the VM (optimizer) determines that no code-path passing [key] to an FFI
-/// function can be called again. For this reason, users should take extra care
-/// to make sure that all accesses to [key] takes an extra reference.
-void _attachFinalizerEVP_PKEY(ffi.Pointer<EVP_PKEY> key) {
-  final ret = dl.webcrypto_dart_dl_attach_finalizer(
-    key,
-    key.cast(),
-    ssl.addresses.EVP_PKEY_free.cast(),
-    // We don't really have an estimate of how much space the EVP_PKEY structure
-    // takes up, but if we make it some non-trivial size then hopefully the GC
-    // will prioritize freeing them.
-    4096,
-  );
-  if (ret != 1) {
-    throw AssertionError('package:webcrypto failed to attached finalizer');
+/// The [_finalizerReachabilityFence] ensures that the [_EvpPKey] is not garbage
+/// collected while the wrapped key is use. Thus, we can be sure the finalizer
+/// won't be called while the wrapped key is in use.
+class _EvpPKey {
+  final ffi.Pointer<EVP_PKEY> _pkey;
+
+  /// Allocate new [EVP_PKEY], attach finalizer and return the wrapped key.
+  factory _EvpPKey() {
+    final _pkey = ssl.EVP_PKEY_new();
+    _checkOp(_pkey.address != 0, fallback: 'allocation failure');
+    return _EvpPKey.wrap(_pkey);
+  }
+
+  /// Wrap existing [EVP_PKEY], this will attach a finalizer.
+  ///
+  /// After this, the wrapped key may only be used within a callback passed to
+  /// [use]. Otherwise, the garbage collect may be calling the finalizer while
+  /// the key is in use.
+  _EvpPKey.wrap(this._pkey) {
+    final ret = dl.webcrypto_dart_dl_attach_finalizer(
+      this,
+      _pkey.cast(),
+      ssl.addresses.EVP_PKEY_free.cast(),
+      // We don't really have an estimate of how much space the EVP_PKEY structure
+      // takes up, but if we make it some non-trivial size then hopefully the GC
+      // will prioritize freeing them.
+      4096,
+    );
+    if (ret != 1) {
+      throw AssertionError('package:webcrypto failed to attached finalizer');
+    }
+  }
+
+  /// Use the wrapped [EVP_PKEY] in callback [fn].
+  ///
+  /// Note. [fn] is not allowed to return a [Future].
+  T use<T>(T Function(ffi.Pointer<EVP_PKEY> pkey) fn) {
+    try {
+      return fn(_pkey);
+    } finally {
+      _finalizerReachabilityFence(this);
+    }
   }
 }
 
-/// Create an [ssl.EVP_PKEY] with finalizer attached.
-///
-/// See [_attachFinalizerEVP_PKEY] for notes on how the finalizer works.
-ffi.Pointer<EVP_PKEY> _createEVP_PKEYwithFinalizer() {
-  final key = ssl.EVP_PKEY_new();
-  _checkOp(key.address != 0, fallback: 'allocation failure');
-  _attachFinalizerEVP_PKEY(key);
-  return key;
+/// Extension of native function that takes a [EVP_PKEY], making it easy to call
+/// using a wrapped [_EvpPKey].
```

Comment:
Instead of this pattern with `use`. Would it be possible to implement them as methods on `_EvpPKey`?

For now, you could add the `reachabilityFence(this)` at the end (and the returns) of those methods. But when `Finalizable` lands, `this` will always be kept alive. See the doc-comment on `Finalizable` in https://dart-review.googlesource.com/c/sdk/+/227501/16/sdk/lib/ffi/native_finalizer.dart.

Reply:
I could, but the `use` method allows me to make extensions like:
```dart
extension<T, A1> on T Function(ffi.Pointer<EVP_PKEY>, A1) {
  /// Invoke this function with unwrapped [key].
  T invoke(_EvpPKey key, A1 arg1) => key.use((pkey) => this(pkey, arg1));
}
```

So then I can write: `ssl.EVP_marshal_private_key.invoke(cbb, _key)`

Instead of having to re-type the signature for `EVP_marshal_private_key` on `_EvpPKey.EVP_marshal_private_key`.

I think I would have to re-type the signature for 10-12 methods... which is kind of annoying, since these methods are generated.

hmm, I guess in some future we can dream about doing something magical with ffigen... hmm... maybe we only dream...

Reply:
@dcharkes I'm curious, with `Finalizable` what are you going to do if a function requires two arguments that are both wrapped in `Finalizable`?

Do you then have to do double dispatch? And if you don't have a dummy `use` method like mine... will you then have double dispatch for each method.

And what if there is 3 `Finalizable` arguments? do we need triple dispatch? :see_no_evil:

Reply:
Any variable stays alive (including this, params and normal variables). For more details see the design doc and language proposal.

Reply:
> I think I would have to re-type the signature for 10-12 methods

Unless all the other things also have Dart types wrapping them, exposing some object oriented API.

Reply:
Hm, it looks like you have 3 layers of objects: `ffigen`erated objects, then `Finalizable` objects, and then a public API. I was thinking the public API would be `Finalizable`s. But it doesn't look like we have a candidate public API type to make finalizable instead of `_EvpPKey`.

Reply:
We do `class RsaPssPublicKey extends _EvpPKey`, and do similar for all the key-classes.. I actually did that in one iteration of this, but then I have to pass a constructor-tearoff to helper methods like `_importPkcs8RsaPrivateKey`.

And the whole design got a lot more complicated. If I do that refactoring, I can do a bit more code deduplication.
But it's a larger refactor, so I opted to do that another day. It also wasn't clear that there would be much win to doing it.

But, yes, if `RsaPssPublicKey extends Finalizable`, then I probably won't need to do anything wrapping methods like `EVP_marshal_private_key` at all.

Reply:
@jonasfj would you like to postpone the refactor? Then I will review the pattern in this CL for now.

Reply:
I think it'll be easier to review if we do one thing at the time.

This PR changes:
 * All occurences of `ffi.Pointer<EVP_PKEY>` to `_EvpPKey`
 * All methods using `ffi.Pointer<EVP_PKEY>` from `ssl.xxx(key, ...)` to `ssl.xxx.invoke(key, ...)`.
, where `.invoke` is a extension on the function type that calls `_EvpPKey.use`.

That's pretty mechanical... if we migrate to put the finalizer on `RsaPssPublicKey` (and friends), then I think it'll take a lot longer to review.

When we get Dart 2.17, it makes sense to migrate the finalizer to sit on `RsaPssPublicKey` (and friends), and then I won't need the _reachability fence_ and I won't need to do anything magic when calling functions that takes a `ffi.Pointer<EVP_PKEY>`.